src/safeposix/syscalls/fs_calls.rs

```diff
@@ -4111,60 +4144,121 @@ impl Cage {
+                    
                     if let Some(segment) = metadata.shmtable.get_mut(&shmid) {
                         for cageid in segment.attached_cages.clone().into_read_only().keys() {
-                            // iterate through all cages with segment attached and add semaphor in segments at attached addr + offset
+                            // iterate through all cages containing segment
                             let cage = interface::cagetable_getref(*cageid);
+                            // Find all addresses in the shared memory region that belong to the current segment.
                             let addrs = Self::rev_shm_find_addrs_by_shmid(&rev_shm, shmid);
+                            // Iterate through all addresses and add the semaphore to the cage's semaphore table.
                             for addr in addrs.iter() {
                                 cage.sem_table.insert(addr + offset, new_semaphore.clone());
                             }
                         }
+                        // Add the offset to the semaphore offsets list.
                         segment.semaphor_offsets.insert(offset);
                     }
                 }
             }
             return 0;
         }
-
+        // Return an error indicating that the semaphore is already initialized.
         return syscall_error(Errno::EBADF, "sem_init", "semaphore already initialized");
     }
 
+/// ## `sem_wait_syscall`
+///
+/// ### Description
+/// 1. Check for Semaphore Existence:The function first checks if the provided 
+/// semaphore handle exists in the semaphore table.
+/// 2. Acquire Semaphore: If the semaphore exists, the function attempts to acquire it using `lock`. 
+/// This operation will block the calling process until the semaphore becomes available.
+/// 3. Error Handling:If the semaphore handle is invalid, the function returns an error
+/// 4. This function allows a process to wait for a semaphore to become available. 
+/// If the semaphore is currently available (its value is greater than 0), the function will 
+/// acquire the semaphore and return 0.
+/// 5. If the semaphore is unavailable (its value is 0), the function will block the 
+/// calling process until the semaphore becomes available(its value becomes 1).
+///[sem_wait(2)](https://man7.org/linux/man-pages/man3/sem_wait.3.html)
+/// ### Function Arguments
+/// * `sem_handle`: A unique identifier for the semaphore.
+///
+/// ### Returns
+/// * 0 on success.
+/// ### Errors
+/// * `EINVAL(22)`: If the semaphore handle is invalid.
+/// * 'EAGAIN(11)' & 'EINTR(4)' currently are not supported 
     pub fn sem_wait_syscall(&self, sem_handle: u32) -> i32 {
         let semtable = &self.sem_table;
         // Check whether semaphore exists
         if let Some(sementry) = semtable.get_mut(&sem_handle) {
+            // Clone the semaphore entry to avoid modifying the original entry in the table.
```

Comment:
I don't think this is actually what is happening with clone here. This is actually dealing with DashMap concurrency so the map isn't blocking.